text/0001-cdk-update.md

```diff
@@ -0,0 +1,293 @@
+---
+rfc pr: [#335](https://github.com/aws/aws-cdk-rfcs/pull/335)
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/1
+---
+
+# CDK Update
+
+Rapid and automatic updates for code-only changes to CDK application stacks.
+
+## Working Backwards
+
+### CHANGELOG
+
+- feat(cli): implement `cdk update` and `cdk update --watch` for asset updates
+
+### README
+
+The `cdk update` command accelerates the asset deployments for your CDK
```

Comment:
I think this needs to be part of the `deploy` command and not a separate command. Eventually. this is about deployment of CDK apps. It's not a separate operation.

If I work backwards from the user experience, I don't think users need to "care" if they are short-circuiting or not. They just want to develop their app and have the CLI deploy it continuously and as fast as possible.

So, primarily this feature should be supported via a `--watch` or (`-w`) switch to `cdk deploy`:

```shell
cdk deploy --watch
```

I expect this command to synth & deploy your app and then watch for changes, synthesize your `cdk.out` directory and perform deployments as needed (with or without short-circuiting). In case the deployment only includes resources that can be short-circuited, it would run the short-circuiting logic. Otherwise, it will perform a normal deployment.

We can add a switch called `--fast` or `--short-circuit` (or something else) which will control the behavior of this feature. It has three modes: `auto`, `never` and `only`.

1. `--fast=auto` - if the update includes only short-circuitable resources, it will apply them. If the update includes non-short-circuitable resources, it will perform a normal  deployment. This is the default if `--watch` is enabled.
2. `--fast=never` - this is the default for `cdk deploy` - never perform a short-circuit deployment.
3. `--fast=only` - only perform short-circuit deployments - fails if there are other updates.

Here are some examples:

```shell
cdk deploy --fast=never  # equivalent to "cdk deploy"
cdk deploy --fast=auto  # one-off, allows both normal and short-circuit updates
cdk deploy --fast=only   # one-off, only allows short-cirtcuitable updates
cdk deploy --watch --fast=auto    # continuous, equivalent to "cdk deploy --watch"
cdk deploy --watch --fast=only     # continuous, only allows short-circuitable updates
cdk deploy --watch --fast=never   # continuous, will perform normal CFN updates
```

Reply:
Here's where we differ; I think it's important to separate the deploy operation from this one, because they are _wildly_ different in what they require. `deploy` requires, generally, _CloudFormation_ access and a role that can interact with that set of tools. _`update`_ requires administrative level access to the account, and directly modifies the resources therein. That's different enough that I'm not comfortable, from a UX point of view, trying to bundle them into the same command. It would be confusing to have `deploy` actually need _different credentials_ depending on what flags you are passing it, whereas here we can document a clear boundary: When you are using `deploy`, you are acting in the role of the production deployment of the system, and when you are using `update` (name tbd :) ) you are acting in the role of the developer.

I would rather see more commands than fewer, and with a more clearly delineated set of requirements, capabilities, and responsibilities.

I think that the ability to perform a fast update vs a full deploy is useful -- the `--fast=auto` you mention above.

Reply:
The CDK CLI commands are designed around user intent, not around their implementation. The "deploy" command includes asset publishing, diff checks, change set creation, etc. From a user standpoint, they just want to deploy.

Same here, the intent of the developer using this feature is to deploy their development stack and therefore this feature is a sub-feature of the "deploy" command and not a separate command.

> I would rather see more commands than fewer, and with a more clearly delineated set of requirements, capabilities, and responsibilities.

I don't think "more" or "less" is the guiding design principle. Our primary concern when we design APIs is the mental model of the user and their intent, and we want to do the "right thing" with the minimum amount of required cognitive load or leakage of the implementation.

I'll frame it differently - if CFN updates would have been very fast and we didn't need to implement these short-circuits (i.e. `cdk deploy` would always take 1s), do you think we would need a separate command to deploy stacks? The answer is _no_. Our users "don't care" that CFN updates take a long time - they just want to deploy their apps and go on with their lives. This is the experience we are striving for (and perhaps at some point CFN will be fast enough that we wouldn't need those short circuits).

Reply:
I think your counterfactual is understandable, but I think we need to put the user's expectations front and centre here, and one of the things that having a separate command does is put a very bright line around the behaviour, which is _wildly_ different from the `deploy` capability. Doing it in the `deploy` command is _possible_, sure; I'm not going to die on that hill, but I can tell you in my experience using CLIs, when you have a command that dramatically changes its mode of operation when some flags are changed -- not just by altering details, but a completely new mode! -- it is surprising and hard to reason about. "Are there combinations of flags that can cause this? What if there are other commands that have similar flags? Do those do the same thing?" A flag introduces the possibility of a cross-cutting function.

I don't believe this is the same thing as a deployment; I don't think we should put it in the same place.